File: python/top_k_frequent_elements.py
from collections import defaultdict
import heapq


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        """time complexity: O(n) where n is the length of nums"""
        result: list[int] = []
        occur_map: dict[int, int] = {}
        bucket: list[list[int]] = []
        for i in range(len(nums) + 1):
            bucket.append([])
        for num in nums:
            if num not in occur_map:
                occur_map[num] = 1
            else:
                occur_map[num] += 1
        for key, val in occur_map.items():
            bucket[val].append(key)
        i: int = 0
        while k > 0 and i < len(nums):
            idx: int = len(nums) - i
            if len(bucket[idx]) > 0:
                result += bucket[idx]
                k -= len(bucket[idx])
            i += 1
        return result

    def topKFrequent2(self, nums: list[int], k: int) -> list[int]:
        """time complexity: O(nlogk) where n is the length of nums, k is the number of unique elements"""
        result: list[int] = []
        count_map: dict[int, int] = defaultdict(int)
        for num in nums:
            count_map[num] += 1
        min_heap: list[tuple[int, int]] = []
        for key, val in count_map.items():
            heapq.heappush(min_heap, (val, key))
        for _ in range(len(min_heap) - k):
            heapq.heappop(min_heap)
        while len(min_heap) > 0:
            count, num = heapq.heappop(min_heap)
            result.append(num)
        return result


def test_topKFrequent_case_1():
    # Arrange
    nums: list[int] = [1, 1, 1, 2, 2, 3]
    k: int = 2
    expected: list[int] = [1, 2]

    # Act
    actual: list[int] = Solution().topKFrequent(nums, k)

    # Assert
    assert expected == actual


def test_topKFrequent_case_2():
    # Arrange
    nums: list[int] = [1]
    k: int = 1
    expected: list[int] = [1]

    # Act
    actual: list[int] = Solution().topKFrequent(nums, k)

    # Assert
    assert expected == actual


def test_topKFrequent_case_3():
    # Arrange
    nums: list[int] = [4, 1, -1, 2, -1, 2, 3]
    k: int = 2
    expected: list[int] = [-1, 2]

    # Act
    actual: list[int] = Solution().topKFrequent(nums, k)

    # Assert
    assert expected == actual

File: rust/src/code/top_k_frequent_elements.rs
use std::collections::HashMap;

pub struct Solution {}

impl Solution {
    // NOTE: Another way can work
    // pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
    //     let mut result: Vec<i32> = Vec::new();
    //     let mut occur_map: HashMap<i32, i32> = HashMap::new();
    //     let mut bucket: Vec<Vec<i32>> = Vec::new();
    //     for _ in 0..nums.len() + 1 {
    //         bucket.push(Vec::new());
    //     }
    //     for num in nums.iter() {
    //         occur_map.entry(*num).and_modify(|x| *x += 1).or_insert(1);
    //     }
    //     for (key, val) in occur_map.iter() {
    //         bucket[*val as usize].push(*key);
    //     }
    //     let mut i: usize = 0;
    //     let mut k: i32 = k;
    //     while k > 0 && i < nums.len() {
    //         let idx: usize = nums.len() - i;
    //         if bucket[idx].len() > 0 {
    //             k -= bucket[idx].len() as i32;
    //             result.append(&mut bucket[idx]);
    //         }
    //         i += 1;
    //     }
    //     result
    // }

    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut map: HashMap<i32, i32> = HashMap::new();
        let mut bucket: Vec<Vec<i32>> = vec![Vec::new(); nums.len() + 1];
        let mut res: Vec<i32> = Vec::new();

        nums.iter().for_each(|x| {
            map.entry(*x)
                .and_modify(|counter| *counter += 1)
                .or_insert(1);
        });

        map.iter().for_each(|(k, v)| {
            bucket[*v as usize].push(*k);
        });

        let mut count: i32 = k;
        let mut i: usize = 0;
        loop {
            if count <= 0 || i > nums.len() {
                break;
            }

            let index: usize = nums.len() - i;
            if bucket[index].len() != 0 {
                count -= bucket[index].len() as i32;
                res.append(&mut bucket[index]);
            }
            i += 1;
        }

        return res;
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn top_k_frequent_case_1() {
        // arrange
        let input = vec![1, 1, 1, 2, 2, 3];
        let k = 2;
        let expected = vec![1, 2];

        // act
        let mut actual = Solution::top_k_frequent(input, k);

        // assert
        actual.sort();
        assert_eq!(actual, expected);
    }

    #[test]
    fn top_k_frequent_case_2() {
        // arrange
        let input = vec![4, 1, -1, 2, -1, 2, 3];
        let k = 2;
        let expected = vec![-1, 2];

        // act
        let mut actual = Solution::top_k_frequent(input, k);

        // assert
        actual.sort();
        assert_eq!(actual, expected);
    }
}
